lib/tpm2_util.c

```diff
@@ -1100,11 +1102,142 @@ TPMI_ALG_HASH *param_hash_algorithm, tpm2_session **session) {
+            continue;
+        }
+
+        tpm2_session *session = NULL;
+        tool_rc rc = tpm2_session_restore(ectx,
+        adjusted_session_paths[session_idx], false, &session);
+        if (rc != tool_rc_success) {
+            LOG_WARN("Could not restore session #%s",
+            adjusted_session_paths[session_idx]);
+            tpm2_session_close(&session);
+            continue;
+        }
+
+        TPM2_SE session_type = tpm2_session_get_type(session);
+        if (session_type != TPM2_SE_HMAC && session_type != TPM2_SE_POLICY) {
+            tpm2_session_close(&session);
+            continue;
+        }
+
+        rethash = tpm2_session_get_authhash(session);
+
+        TPMA_SESSION attrs = 0;
+        ESYS_TR session_handle = tpm2_session_get_handle(session);
+        tool_rc tmp_rc = tpm2_sess_get_attributes(ectx, session_handle, &attrs);
+        UNUSED(tmp_rc);
+        if (attrs & TPMA_SESSION_AUDIT) {
+            tpm2_session_close(&session);
+            break;
+        }
+    }
+
+    return rethash;
+}
+
+TPMI_ALG_HASH calc_cphash_alg_from_cphash_path(const char *cphash_path) {
+
+    if (!cphash_path) {
+        return TPM2_ALG_ERROR;
+    }
+
     /*
-     * Parameter hash alg is setup as that of the session alg of first session
+     * Expecting single token, so tokenize just once.
      */
-    if (param_hash_algorithm) {
-        *param_hash_algorithm = tpm2_session_get_authhash(session[0]);
+    char *str = (char *)cphash_path;
+    char *token = strtok(str, ":");
+    if (!token) {
+        return TPM2_ALG_ERROR;
     }
-    return tool_rc_success;
+
+    TPMI_ALG_HASH hashalg = tpm2_alg_util_from_optarg(token,
+    tpm2_alg_util_flags_hash);
+    if (hashalg == TPM2_ALG_ERROR) {
+        LOG_WARN("Invalid/ Unspecified policy digest algorithm.");
+    }
+
+    return hashalg;
 }
+
+/*
+ * It should be noted that the auths aren't checked when calculating the cpHash,
+ * instead the sessions are consumed to determine the cpHash algorithm.
+ *
+ * 1. If cphash_path is preceded with <halg>: use that as cphash-halg return
+ * Otherwise
+ *
+ *  Consume session only if it is a policy-session or an hmac-session
+ *  1. If only hmac or policy session is specified, return that session's halg
+ *  2. If hmac-session with audit is specified, return that session's halg
+ *  3. If policy-session, then return policy-session's halg
+ *
+ * Otherwise
+ *  return SHA256
+ */
+TPMI_ALG_HASH tpm2_util_calculate_cphash_alg(ESYS_CONTEXT *ectx,
+const char *cphash_path, const char **sessions_auths_or_paths) {
+
+    TPMI_ALG_HASH cphash_alg = calc_cphash_alg_from_cphash_path(cphash_path);
+    if (cphash_alg != TPM2_ALG_ERROR) {
+        return cphash_alg;
+    }
+
+    bool is_sessions_path_valid = adjust_sessions_auths_or_paths(
+        sessions_auths_or_paths);
+    if (!is_sessions_path_valid) {
+        LOG_WARN("Invalid session count or paths: Defaulting cphash alg to sha256");
+        return TPM2_ALG_SHA256;
```

Comment:
Is this user input? If user specifies input, shouldn't we fail?

Reply:
I guess is_sessions_path_valid could be better phrased and also remove the warning to make the outputs less noisy :). It is possible that user correctly specifies password-only sessions and in those cases we default to sha256.